From 13f01d5f7b7365286fe819a6c9c9e169dd06d602 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Fri, 7 Jul 2017 01:40:53 +0100
Subject: [PATCH] Ensure we allow launching .desktop files with trusted symlink
 targets

When a file is a symlink to one of the XDG data dirs, we'll allow that
link to work, as they're vendor provided and not world-writeable by a
malicious entity.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 src/nautilus-directory-async.c | 42 ++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 42 insertions(+)

diff --git a/src/nautilus-directory-async.c b/src/nautilus-directory-async.c
index 05e138a..ed25123 100644
--- a/src/nautilus-directory-async.c
+++ b/src/nautilus-directory-async.c
@@ -3575,6 +3575,43 @@ file_info_start (NautilusDirectory *directory,
     g_object_unref (location);
 }
 
+static gboolean is_trusted_system_desktop_file (GFile *file)
+{
+    gboolean res = FALSE;
+    GFileInfo *info;
+    const gchar *target = NULL;
+    GFile *location = NULL;
+
+    info = g_file_query_info (file,
+                              G_FILE_ATTRIBUTE_STANDARD_TYPE ","
+                              G_FILE_ATTRIBUTE_STANDARD_SYMLINK_TARGET,
+                              G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS,
+                              NULL,
+                              NULL);
+
+    if (info == NULL)
+    {
+        return FALSE;
+    }
+
+    target = g_file_info_get_symlink_target (info);
+    if (!target) {
+        goto done;
+    }
+
+    location = g_file_new_for_path (target);
+
+    res = nautilus_is_in_system_dir (location);
+
+done:
+    if (location) {
+        g_object_unref (location);
+    }
+    g_object_unref (info);
+
+    return res;
+}
+
 static gboolean
 is_link_trusted (NautilusFile *file,
                  gboolean      is_launcher)
@@ -3602,6 +3639,11 @@ is_link_trusted (NautilusFile *file,
     {
         location = nautilus_file_get_location (file);
         res = nautilus_is_in_system_dir (location);
+
+        if (!res) {
+            res = is_trusted_system_desktop_file (location);
+        }
+
         g_object_unref (location);
     }
 
-- 
2.13.2

